### PR TITLE
 Integrate DPDK capture interface into Suricata v2

### PR DESCRIPTION
**Pull request by lukashino**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/issues/4507) ticket:

Describe changes:
- Add STAILQ implementation (required for DPDK).
- Integrate DPDK capture interface as an alternative to AF_PACKET or others.
- Focus implementation on workers runmode.
- Support both IDS and IPS modes.
- Follow AF_PACKET capture interface architecture and apply DPDK best practices for development.
- Update documentation to reflect new changes.

Work has been based on DPDK 21.05 but I believe running it on earlier versions would be also possible.

This is a follow-up pull request of #6159. I have reformatted all source file changes according to clang. However the check will probably not pass regardless because there are still many source files not following clang style guides. Please correct me if I am wrong.

### CODE REVIEW

**Comment by lukashino** [comment]
Hi @victorjulien, I think I have implemented all requirements for Suricata pull request. Would it be please possible that somebody from the Suricata team could have a look at my PR? Thank you.

File: src/queue.h [hunk 1]
```diff
@@ -401,120 +401,118 @@ struct {								\
 /*
  * Tail queue definitions.
  */
-#define TAILQ_HEAD(name, type)						\
```
Comment (catenacyber) [inline src/queue.h:321]: Why do we need these formatting changes ?

File: src/runmodes.c [hunk 2]
```diff
@@ -157,6 +156,13 @@ static const char *RunModeTranslateModeToName(int runmode)
 #else
             return "WINDIVERT(DISABLED)";
 #endif
+#ifdef HAVE_DPDK
```
Comment (catenacyber) [inline src/runmodes.c:159]: `ifdef` should come after `case` as for WINDIVERT for instance

File: src/suricata-common.h [hunk 3]
```diff
@@ -40,6 +40,11 @@
 #include <autoconf.h>
 #endif
 
+#ifdef HAVE_DPDK
+#define HAVE_STRLCAT 1
+#define HAVE_STRLCPY 1
```
Comment (catenacyber) [inline src/suricata-common.h:45]: Why is this here ?
Should not these `HAVE_STRLCPY` be set by configure in config.h ?

File: src/tm-threads.c [hunk 4]
```diff
@@ -248,7 +253,14 @@ static void *TmThreadsSlotPktAcqLoop(void *td)
                                  " tmqh_out=%p",
                    s, s ? s->PktAcqLoop : NULL, tv->tmqh_in, tv->tmqh_out);
         TmThreadsSetFlag(tv, THV_CLOSED | THV_RUNNING_DONE);
-        pthread_exit((void *) -1);
+#ifdef HAVE_DPDK
+        if (tv->lcore_id == LCORE_ID_ANY) {
```
Comment (catenacyber) [inline src/tm-threads.c:257]: What is happening here ? I think this deserves a comment

File: .github/workflows/builds.yml [hunk 5]
```diff
@@ -185,6 +185,7 @@ jobs:
                 automake \
                 cargo-vendor \
                 diffutils \
+                dpdk \
```
Comment (catenacyber) [inline .github/workflows/builds.yml:188]: I think some configure should also add `--enable-dpdk`, right ?

**Review by catenacyber — COMMENTED** [review]
Thanks for this work.
I left a few comments to allow you to continue working on this even if I do not know everything about DPDK

**Comment by catenacyber** [comment]
Formatting check is off:
You can fix it with `git rebase master -x "git clang-format HEAD^"`

**Comment by lukashino** [comment]
> Formatting check is off:
> You can fix it with `git rebase master -x "git clang-format HEAD^"`

Thank you for taking your time to review the code. I will apply the command you posted and then work on the rest.

File: src/runmodes.c (hunk 2 above)
Comment (lukashino) [reply src/runmodes.c:159]: Fixed, thanks!

File: src/suricata-common.h (hunk 3 above)
Comment (lukashino) [reply src/suricata-common.h:45]: Fixed, I moved STRLCAT/STRLCPY to configure.ac

File: .github/workflows/builds.yml (hunk 5 above)
Comment (lukashino) [reply .github/workflows/builds.yml:188]: Good catch, added the switch to CentOS8 build.

File: src/tm-threads.c (hunk 4 above)
Comment (lukashino) [reply src/tm-threads.c:257]: In #6269 I have tried to be more descriptive at all places where LCORE_ID_ANY is used. I also did a bit of refactoring. Thanks for pointing this out.

**Comment by lukashino** [comment]
Pull request #6269 is the successor of the current pull request. Therefore, I am closing current pull request.
